design/design-doc-observe-only-resources.md

```diff
@@ -0,0 +1,293 @@
+`spec.managementPolicy` when the feature is not enabled.
+
+#### Deprecation of `deletionPolicy`
+
+With the new `managementPolicy` covering the existing `deletionPolicy`, we will deprecate the latter in favor of the
+former.
+
+Until we drop the `deletionPolicy` from the schema altogether, we need to be careful with the conflicting combinations
+shown in the below table which only exists with deletion:
+
+| Deletion Policy | Management Policy | Should Observe? | Should Create? | Should Update? | Should Delete? |
+| --- | --- | --- | --- | --- | --- |
+| Delete | Full | Yes | Yes | Yes | Yes |
+| Orphan | ObserveCreateUpdate | Yes | Yes | Yes | No |
+| Delete | ObserveOnly | Yes | No | No | Conflict (No) |
+| Orphan | Full | Yes | Yes | Yes | Conflict (No) |
+| Delete | ObserveCreateUpdate | Yes | Yes | Yes | Conflict (No) |
+| Orphan | ObserveOnly | Yes | No | No | No |
+
+For conflicting cases, we will decide based on the non-default configuration and not delete the external resource.
+This way, we will also err on the side of caution by leaving the actual resource. However, this comes at the cost of
+some possible confusion for users focusing on only one of the policies.
+
+Another solution could be simply throwing an error and preventing reconciliation during conflict. This would be more
+explicit but would require some manual actions and degraded UX for the usage of the feature, for example:
+
+- Creating an ObserveOnly resource will require both setting `managementPolicy` to `ObserveOnly` and `deletionPolicy`
+to `Orphan` .
+- If there are existing resources with `deletionPolicy: Orphan` when the feature is enabled, they will start failing to
+reconcile until their `managementPolicy`’s updated to `ObserveCreateUpdate`.
+
+#### Schema Changes
+
+The proposed approach here involves utilizing the same CR, hence schema, for both managing and observing resources.
+The caveat here is that some fields are required for creating the resources but not for observing. This means that it
+won’t be possible to create an observe-only resource without providing a value for any required fields, because
+the Kubernetes API checks for the presence of these fields before allowing the CR to be created.
+
+> Please note this is already the case with [importing existing managed resources] today.
+
+We will fix this by leveraging the [Common Expression Language (CEL)], which was graduated to beta (i.e. enabled by default)
+as of Kubernetes 1.25. See the following diff for the required changes we need for making [CIDRBlock] parameter of
+AWS VPC required only if not `ObserveOnly`:
+
+```diff
+		// CIDRBlock is the IPv4 network range for the VPC, in CIDR notation. For
+        // example, 10.0.0.0/16.
+-       // +kubebuilder:validation:Required
+        // +immutable
+-       CIDRBlock string `json:"cidrBlock"`
++       CIDRBlock *string `json:"cidrBlock,omitempty"`
```

Comment:
This appears to indicate that in order to get the most benefit from this feature either k8s 1.25 or higher is required, or the validation rules feature gate needs to be enabled on k8s <= 1.24.  I don't think the feature is nearly as useful if the resource required fields are still required (e.g. VPC cidrBlock, Cluster resourcesVpcConfig).

It also means a lot of work for the (existing, manual) providers to rework the resources with these changes.

Does upjet provide a way to implement this automatically across all resources?  That may be another reason to move to the generated providers if they automatically support observe-only resources.

Reply:
> This appears to indicate that in order to get the most benefit from this feature either k8s 1.25 or higher is required, or the validation rules feature gate needs to be enabled on k8s <= 1.24. I don't think the feature is nearly as useful if the resource required fields are still required (e.g. VPC cidrBlock, Cluster resourcesVpcConfig).

Once we do this change here for a given resource, Observe Only resources would immediately work without needing all required fields to be provided, whether CEL is available or not. What users with k8s <= 1.24 (without the alpha feature enabled) lose is that when, management policy is not ObserveOnly, they will no longer get a validation error at the API server level, during CR creation when they didn't provide a required field.

This would definitely be a UX degradation; however, it shouldn't be a huge deal assuming the reconciler will report the error from cloud API indicating that a required field is missing. 

The question is whether we can live with it until k8s 1.25 is ubiquitous or if should we consider providing a temporary solution like a validation webhook that understands CEL and apply the validation there?

 > It also means a lot of work for the (existing, manual) providers to rework the resources with these changes.

We don't have to add this support to all (manual) resources at once, but yes, all resource schemas needs to be updated accordingly.

> Does upjet provide a way to implement this automatically across all resources? That may be another reason to move to the generated providers if they automatically support observe-only resources.

Yes, I believe it would be straightforward to automate this with upjet, and we can it available for all resources at once.

Reply:
> it shouldn't be a huge deal assuming the reconciler will report the error from cloud API indicating that a required field is missing.

this sounds reasonable to me honestly - my initial concern was that folks wouldn't be able to run Crossplane on < 1.25 at all (i.e. Crossplane was taking a hard dependency on 1.25+), which would definitely be undesirable :)